### ISSUE TITLE
JSON.stringify does not allow returning undefined

### ISSUE THREAD

**Issue by kryptt** [scala-js/scala-js#4094]
Calling JSON.stringify with undefined, or a function returns undefined in most contexts (browsers)

```
val x = scala.scalajs.js.JSON.stringify(())

println(s"found: $x" )
```
see: https://scalafiddle.io/sf/oFq7bkS/0
to reproduce.

I believe there are two directions to go, either change the return type to UndefOr[String] or document that this function may throw an exception.

**Comment by ritschwumm** [comment]
if it returns undefined in JS, then it should return undefined in scala-js.

standard apis are always described, never extended or changed in behaviour, which is a good thing (tm).

**Comment by kryptt** [comment]
You will notice if you run the fiddle, it does not just return undefined. It will throw a ClassCastException. 

In fact I realized the behavior should be different betwee fastOptJS and fullOptJS. For fastOptJS it actually gets wrapped: `$as_T(JSON.stringify(value))`.

This should probably be treated as a bug.

**Comment by gzm0** [comment]
> if it returns undefined in JS, then it should return undefined in scala-js.

@ritschwumm It you refer to the facade definition, this is not necessarily true. For example `js.Array[T]`'s index access (i.e. `a[2]`) specifies a return type of `T`. However, if you access a JavaScript array out of bounds, it'll return `undefined`.

It all boils down to defining what "valid" usages of a certain API are. [MDN says about `JSON.stringify`][1]:

> undefined, Functions, and Symbols are not valid JSON values. If any such values are encountered during conversion they are either omitted (when found in an object) or changed to null (when found in an array). JSON.stringify() can return undefined when passing in "pure" values like JSON.stringify(function(){}) or JSON.stringify(undefined).

To me, this clearly indicates that the caller is responsible to ensure an allowed value is passed to `JSON.stringify`.

As such, IMHO the only thing to do here is to update the documentation.

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify

**Comment by kryptt** [comment]
@gzm0  How do you normally handle discrepancies between fastOptJS and fullOptJS ?

**Comment by gzm0** [comment]
@kryptt Have a look at the section "Undefined Behavior" here: https://www.scala-js.org/doc/semantics.html

tl;dr: You do not (after the fact). If it triggers, your code has a bug you need to fix.

In the case of `JSON.stringify`: Don't pass a thing that is not a valid JSON value. In the case of ` js.Array`: Do not access out of bounds.

**Comment by sjrd** [comment]
I agree with @gzm0 here. At the end of the day, it is a bit of a judgment call what uses of an API are considered "valid". Very old JavaScript APIs like `JSON` are *very* permissive, often more for legacy reasons than careful design. If we always type the most permissive interpretation of those APIs, the expected, regular usages become very clunky. It becomes a trade-off.

In some cases like `js.Array`, we make an intentional decision to restrict the facades to "valid" use cases, for some definition of "valid" that we chose. In the case of `JSON.stringify`, I don't remember it being an intentional decision when we added it, but I think that sticking to `String` is reasonable.
